File: src/agents/opencode.ts
import { existsSync, rmSync } from "node:fs";
import { join } from "node:path";
import { getOrCreateObject, hasKey, readJsonFile, writeJsonFile } from "../config/json.js";
import type { Detection, RunOpts, ToolId } from "../registry.js";
import {
  opencodePluginInstalled,
  registerCavemanOpencode,
  unregisterCavemanOpencode,
} from "../tools/caveman.js";
import { verbose } from "../util/colors.js";
import { findBinaryIn, isOnPath } from "../util/detect.js";
import * as paths from "../util/paths.js";
import { hasOwner, removeOwner, writeOwner } from "../util/unified-block.js";

/** Detect if OpenCode is installed. */
export function detect(): Detection {
  const hasCli = !!findBinaryIn("opencode", paths.opencodeKnownBinDirs());
  const hasDesktop = paths.opencodeDesktopPaths().some((p) => existsSync(p));
  if (hasCli && hasDesktop) return { installed: true, source: "cli+desktop" };
  if (hasCli) return { installed: true, source: "cli" };
  if (hasDesktop) return { installed: true, source: "desktop" };
  if (process.env.NODE_ENV === "test" && existsSync(paths.opencodePaths().dir)) {
    return { installed: true, source: "config" };
  }
  return { installed: false, source: "" };
}

/** Wire a tool into OpenCode. */
export async function wire(tool: ToolId, opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "codegraph":
      wireMcp("codegraph", [paths.toksaveAbs(), "runmcp", "codegraph", "serve", "--mcp"], opts);
      if (!opts.dryRun) {
        writeOwner("opencode", "codegraph");
        unwireAutoIndexOpenCode();
      }
      return true;
    case "context-mode":
      wireMcpContextMode(opts);
      if (!opts.dryRun) writeOwner("opencode", "context-mode");
      return true;
    case "caveman":
      if (!opts.dryRun) {
        registerCavemanOpencode();
        writeOwner("opencode", "caveman");
      }
      return true;
    case "rtk":
      if (!opts.dryRun) {
        wireRtkPlugin(opts);
      }
      return true;
    case "ponytail":
      if (!opts.dryRun) {
        registerPonytailPlugin();
        writeOwner("opencode", "ponytail");
      }
      return true;
    case "principles":
      if (!opts.dryRun) writeOwner("opencode", "principles");
      return true;
    default:
      return false;
  }
}

/** Unwire a tool from OpenCode. */
export async function unwire(tool: ToolId, _opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "codegraph":
      removeMcp("codegraph");
      removeOwner("opencode", "codegraph");
      return true;
    case "context-mode":
      removeContextModePlugin();
      removeOwner("opencode", "context-mode");
      return true;
    case "caveman":
      unregisterCavemanOpencode();
      removeOwner("opencode", "caveman");
      return true;
    case "rtk":
      removeRtkPlugin();
      return true;
    case "ponytail":
      unregisterPonytailPlugin();
      removeOwner("opencode", "ponytail");
      removePonytailArtifacts();
      return true;
    case "principles":
      removeOwner("opencode", "principles");
      return true;
    default:
      return false;
  }
}

/** Verify a tool is wired into OpenCode. */
export function verify(tool: ToolId): boolean | null {
  switch (tool) {
    case "codegraph":
      return hasMcp("codegraph");
    case "context-mode":
      return hasContextModePlugin();
    case "caveman":
      return opencodePluginInstalled() || hasOwner("opencode", "caveman");
    case "rtk":
      return isOnPath("rtk") && hasRtkPlugin();
    case "ponytail":
      return hasOwner("opencode", "ponytail") || ponytailPluginInstalled();
    case "principles":
      return hasOwner("opencode", "principles");
    default:
      return null;
  }
}

// ─── MCP wiring ──────────────────────────────────────────────

function wireMcp(toolId: string, command: string[], opts: RunOpts): boolean {
  if (opts.dryRun) return true;

  verbose(`Wiring MCP ${toolId} into OpenCode`, opts.verbose);

  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};

  if (!hasKey(cfg, "$schema")) {
    cfg.$schema = "https://opencode.ai/config.json";
  }

  const mcp = getOrCreateObject(cfg, "mcp");
  const entry = { type: "local", command, enabled: true };
  const existing = mcp[toolId] as
    | { type?: unknown; command?: unknown; enabled?: unknown }
    | undefined;

  if (
    existing?.type === entry.type &&
    Array.isArray(existing.command) &&
    existing.command.length === command.length &&
    existing.command.every((arg, i) => arg === command[i]) &&
    existing.enabled === entry.enabled
  ) {
    return true;
  }

  mcp[toolId] = entry;
  writeJsonFile(p.config, cfg);
  return true;
}

function wireMcpContextMode(opts: RunOpts): boolean {
  if (opts.dryRun) return true;
  const p = paths.opencodePaths();
  paths.ensureDir(p.dir);
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  if (!hasKey(cfg, "$schema")) cfg.$schema = "https://opencode.ai/config.json";
  // Ensure bare context-mode plugin entry, remove versioned
  setContextModePluginBare(cfg);
  // Also ensure mcp entry for context-mode via local runmcp (if using plugin bare, mcp comes from plugin)
  // For safety, ensure mcp entry not duplicated as old style
  const mcp = getOrCreateObject(cfg, "mcp");
  if (mcp["context-mode"]) {
    delete mcp["context-mode"];
  }
  writeJsonFile(p.config, cfg);
  return true;
}

function setContextModePluginBare(cfg: Record<string, unknown>): void {
  const plugins = (cfg.plugin as unknown[]) ?? [];
  const kept: unknown[] = [];
  for (const p of plugins) {
    if (typeof p === "string" && (p === "context-mode" || p.startsWith("context-mode@"))) continue;
    kept.push(p);
  }
  kept.push("context-mode");
  cfg.plugin = kept;
  // Remove old mcp.context-mode if exists
  const mcp = (cfg.mcp as Record<string, unknown>) ?? {};
  if (mcp["context-mode"]) {
    delete mcp["context-mode"];
    if (Object.keys(mcp).length === 0) delete (cfg as Record<string, unknown>).mcp;
  }
}

function removeMcp(toolId: string): void {
  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  const mcp = cfg.mcp as Record<string, unknown> | undefined;
  if (mcp?.[toolId]) {
    delete mcp[toolId];
    writeJsonFile(p.config, cfg);
  }
}

function hasMcp(toolId: string): boolean {
  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  const mcp = cfg.mcp as Record<string, unknown> | undefined;
  return !!mcp?.[toolId];
}

function hasContextModePlugin(): boolean {
  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  const plugins = cfg.plugin as unknown[] | undefined;
  if (!Array.isArray(plugins)) return false;
  return plugins.some(
    (pl) => typeof pl === "string" && (pl === "context-mode" || pl.startsWith("context-mode@")),
  );
}

function removeContextModePlugin(): void {
  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  const plugins = cfg.plugin as unknown[] | undefined;
  if (!Array.isArray(plugins)) return;
  const kept = plugins.filter(
    (pl) => !(typeof pl === "string" && (pl === "context-mode" || pl.startsWith("context-mode@"))),
  );
  cfg.plugin = kept;
  writeJsonFile(p.config, cfg);
}

// Legacy auto-index cleanup (opencode/plugins/tokless-codegraph-init.js)
function unwireAutoIndexOpenCode(): void {
  const legacyPath = `${paths.opencodePaths().dir}/plugins/tokless-codegraph-init.js`;
  try {
    rmSync(legacyPath, { force: true });
  } catch {}
}

// ─── RTK plugin ──────────────────────────────────────────────

function wireRtkPlugin(opts: RunOpts): void {
  const pluginFile = rtkPluginPath();
  verbose("Installing RTK plugin for OpenCode", opts.verbose);
  paths.writeFile(pluginFile, RTK_PLUGIN);
}

function removeRtkPlugin(): void {
  try {
    rmSync(rtkPluginPath(), { force: true });
  } catch {}
}

function hasRtkPlugin(): boolean {
  return existsSync(rtkPluginPath());
}

function rtkPluginPath(): string {
  return `${paths.opencodePaths().dir}/plugins/toksave-rtk.js`;
}

const RTK_PLUGIN = `export const Plugin = async () => ({
  "tool.execute.before": async (input, output) => {
    if (input.tool !== "bash") return;
    const command = String(output.args.command ?? "").trim();
    if (!command || command === "rtk" || command.startsWith("rtk ")) return;
    output.args.command = \`rtk \${command}\`;
  },
});
`;

// ─── Ponytail plugin ─────────────────────────────────────────

const PONYTAIL_PKG = "@dietrichgebert/ponytail";

function registerPonytailPlugin(): void {
  const p = paths.opencodePaths();
  const cfg = (readJsonFile(p.config) as Record<string, unknown>) ?? {};
  if (!hasKey(cfg, "$schema")) cfg.$schema = "https://opencode.ai/config.json";
  const plugins = (cfg.plugin as unknown[]) ?? [];
  // Check already present
  if (
    plugins.some((pl) => typeof pl === "string" && pl.toLowerCase() === PONYTAIL_PKG.toLowerCase())
  ) {
    return;
  }
  // Insert before context-mode if present
  let inserted = false;
  const out: unknown[] = [];
  for (const pl of plugins) {
    if (
      !inserted &&
      typeof pl === "string" &&
      (pl === "context-mode" || pl.startsWith("context-mode@"))
    ) {
      out.push(PONYTAIL_PKG);
      inserted = true;
    }
    out.push(pl);
  }
  if (!inserted) out.push(PONYTAIL_PKG);
  cfg.plugin = out;
  writeJsonFile(p.config, cfg);
}

function unregisterPonytailPlugin(): void {
  const p = paths.opencodePaths();
  const cfg = readJsonFile(p.config) as Record<string, unknown> | null;
  if (!cfg) return;
  const plugins = cfg.plugin as unknown[] | undefined;
  if (!Array.isArray(plugins)) return;
  const kept = plugins.filter(
    (pl) => !(typeof pl === "string" && pl.toLowerCase() === PONYTAIL_PKG.toLowerCase()),
  );
  cfg.plugin = kept;
  writeJsonFile(p.config, cfg);
}

function ponytailPluginInstalled(): boolean {
  const p = paths.opencodePaths();
  const cfg = readJsonFile(p.config) as Record<string, unknown> | null;
  if (!cfg) return false;
  const plugins = cfg.plugin as unknown[] | undefined;
  if (!Array.isArray(plugins)) return false;
  return plugins.some(
    (pl) => typeof pl === "string" && pl.toLowerCase() === PONYTAIL_PKG.toLowerCase(),
  );
}

function removePonytailArtifacts(): void {
  const dir = paths.opencodePaths().dir;
  try {
    rmSync(`${dir}/plugins/ponytail`, { recursive: true, force: true });
  } catch {}
  try {
    rmSync(`${dir}/.ponytail-active`, { force: true });
  } catch {}
}

// ─── Auto-index plugin ───────────────────────────────────────

const AUTO_INDEX_PLUGIN = `let indexed = false;
export const Plugin = async () => ({
  "tool.execute.before": async () => {
    if (indexed) return;
    indexed = true;
    const { execSync } = require("node:child_process");
    try { execSync("toksave index --auto", { timeout: 120000 }); } catch {}
  },
});
`;

function autoIndexPluginPath(): string {
  return join(paths.opencodePaths().dir, "plugins", "toksave-autoindex.js");
}

export function installOpencodeAutoIndexPlugin(): void {
  paths.writeFile(autoIndexPluginPath(), AUTO_INDEX_PLUGIN);
}

export function removeOpencodeAutoIndexPlugin(): void {
  try { rmSync(autoIndexPluginPath(), { force: true }); } catch {}
}

export function hasOpencodeAutoIndexPlugin(): boolean {
  return existsSync(autoIndexPluginPath());
}

File: src/agents/opencode.rs
use crate::agents::Agent;
use crate::registry::{Detection, RunOpts, ToolId};
use crate::util::detect::find_binary_in;
use crate::util::errors::Result;
use crate::util::json::{
    get_or_create_object, has_key, read_json_file, write_json_file, write_json_pruned,
};
use crate::util::paths::{
    opencode_desktop_paths, opencode_known_bin_dirs, opencode_paths, write_file,
};
use crate::util::unified_block::{has_owner, remove_owner, write_owner};
use std::fs;

/// Agent implementation managing OpenCode integration (v1 and v2).
pub struct OpencodeAgent;

impl OpencodeAgent {
    /// Creates a new `OpencodeAgent` instance.
    pub fn new() -> Self {
        Self
    }
}

impl Default for OpencodeAgent {
    fn default() -> Self {
        Self::new()
    }
}

impl Agent for OpencodeAgent {
    fn detect(&self) -> Detection {
        let p = opencode_paths();
        let has_cli = find_binary_in("opencode", &opencode_known_bin_dirs()).is_some();
        let has_desktop = opencode_desktop_paths().iter().any(|p| p.exists());
        if has_cli && has_desktop {
            return Detection {
                installed: true,
                source: "cli+desktop".to_string(),
            };
        }
        if has_cli {
            return Detection {
                installed: true,
                source: "cli".to_string(),
            };
        }
        if has_desktop {
            return Detection {
                installed: true,
                source: "desktop".to_string(),
            };
        }
        if std::env::var("TOKSAVE_TEST").is_ok() && p.dir.exists() {
            return Detection {
                installed: true,
                source: "config".to_string(),
            };
        }
        Detection {
            installed: false,
            source: String::new(),
        }
    }

    fn wire(&self, tool: ToolId, opts: &RunOpts) -> Result<bool> {
        if opts.dry_run {
            return Ok(true);
        }
        let p = opencode_paths();
        let mut cfg = read_json_file(&p.config)?.unwrap_or_else(|| serde_json::json!({}));
        if !has_key(&cfg, "$schema") {
            cfg["$schema"] = serde_json::json!("https://opencode.ai/config.json");
        }

        match tool {
            ToolId::Codegraph => {
                let mcp = get_or_create_object(&mut cfg, "mcp");
                let server_entry = serde_json::json!({
                    "type": "local",
                    "command": [crate::util::paths::toksave_abs(), "runmcp", "codegraph", "serve", "--mcp"],
                    "enabled": true
                });
                mcp["codegraph"] = server_entry.clone();
                let is_v2 = p.config.file_name().and_then(|f| f.to_str()) == Some("opencode.json")
                    || mcp.get("servers").is_some();
                if is_v2 {
                    let servers = get_or_create_object(mcp, "servers");
                    let mut v2_server = server_entry;
                    v2_server["disabled"] = serde_json::json!(false);
                    servers["codegraph"] = v2_server;
                }
                write_json_file(&p.config, &cfg)?;
                write_owner("opencode", "codegraph")?;
                let plugin_file = p.plugins_dir.join("toksave-autoindex.js");
                let plugin_code = r#"import { execSync } from "node:child_process";
import { Plugin } from "@opencode/plugin";

let indexed = false;
export default Plugin.define({
  id: "toksave.autoindex",
  async setup(ctx) {
    await ctx.tool.hook("execute.before", () => {
      if (indexed) return;
      indexed = true;
      try { execSync("toksave index --auto", { timeout: 120000 }); } catch {}
    });
  },
});
"#;
                write_file(&plugin_file, plugin_code)?;
                Ok(true)
            }
            ToolId::ContextMode => {
                let is_v2 = cfg.get("plugins").is_some()
                    || p.config.file_name().and_then(|f| f.to_str()) == Some("opencode.json");
                let key = if is_v2 { "plugins" } else { "plugin" };
                let plugins = cfg.get_mut(key).and_then(|v| v.as_array_mut());
                let mut plugin_arr = match plugins {
                    Some(arr) => arr.clone(),
                    None => vec![],
                };
                if !plugin_arr.contains(&serde_json::json!("context-mode")) {
                    plugin_arr.push(serde_json::json!("context-mode"));
                }
                cfg[key] = serde_json::Value::Array(plugin_arr);
                write_json_file(&p.config, &cfg)?;
                write_owner("opencode", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                write_owner("opencode", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                let plugin_file = p.plugins_dir.join("toksave-rtk.js");
                // Resolve rtk's absolute path the same way toksave's own hook does: a bare
                // `rtk` prefix fails when OpenCode was launched before rtk's install dir was
                // added to PATH (fresh install, GUI-launched session, stale shell PATH cache).
                let rtk_plugin = r#"import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { Plugin } from "@opencode/plugin";

function resolveRtkBin() {
  let rtkBin = "rtk";
  const localPath = process.platform === "win32"
    ? path.join(process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local"), "Programs", "toksave", "rtk.exe")
    : path.join(os.homedir(), ".local", "bin", "rtk");
  if (fs.existsSync(localPath)) {
    rtkBin = process.platform === "win32" ? localPath.replace(/\//g, "\\") : localPath;
    if (/\s/.test(rtkBin)) rtkBin = `"${rtkBin}"`;
  }
  return rtkBin;
}

export default Plugin.define({
  id: "toksave.rtk",
  async setup(ctx) {
    const rtkBin = resolveRtkBin();
    const alts = [rtkBin, rtkBin.replace(/\\/g, "/"), rtkBin.replace(/\//g, "\\")];
    await ctx.tool.hook("execute.before", (event) => {
      if (event.tool !== "shell" && event.tool !== "bash") return;
      const input = event.input;
      if (!input || typeof input !== "object") return;
      const command = String(input.command ?? "").trim();
      if (!command || /^(rtk|rtk\\.exe)(\\s|$)/.test(command) || alts.some((p) => command === p || command.startsWith(p + " "))) return;
      input.command = `${rtkBin} ${command}`;
    });
  },
});
"#;
                write_file(&plugin_file, rtk_plugin)?;
                Ok(true)
            }
            ToolId::Ponytail => {
                write_owner("opencode", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                write_owner("opencode", "principles")?;
                Ok(true)
            }
        }
    }

    fn unwire(&self, tool: ToolId, _opts: &RunOpts) -> Result<bool> {
        let p = opencode_paths();
        match tool {
            ToolId::Codegraph => {
                if let Some(mut cfg) = read_json_file(&p.config)? {
                    if let Some(mcp) = cfg.get_mut("mcp").and_then(|v| v.as_object_mut()) {
                        mcp.remove("codegraph");
                        if let Some(servers) =
                            mcp.get_mut("servers").and_then(|s| s.as_object_mut())
                        {
                            servers.remove("codegraph");
                        }
                    }
                    write_json_pruned(&p.config, &cfg)?;
                }
                let _ = fs::remove_file(p.plugins_dir.join("toksave-autoindex.js"));
                let _ = fs::remove_dir_all(p.plugins_dir.join("toksave-autoindex"));
                remove_owner("opencode", "codegraph")?;
                Ok(true)
            }
            ToolId::ContextMode => {
                if let Some(mut cfg) = read_json_file(&p.config)? {
                    for key in ["plugin", "plugins"] {
                        if let Some(plugins) = cfg.get_mut(key).and_then(|v| v.as_array_mut()) {
                            plugins.retain(|v| {
                                v.as_str().map(|s| s != "context-mode").unwrap_or(true)
                            });
                        }
                    }
                    if let Some(servers) = cfg
                        .get_mut("mcp")
                        .and_then(|m| m.get_mut("servers"))
                        .and_then(|s| s.as_object_mut())
                    {
                        servers.remove("context-mode");
                    }
                    write_json_pruned(&p.config, &cfg)?;
                }
                remove_owner("opencode", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                remove_owner("opencode", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                let _ = fs::remove_file(p.plugins_dir.join("toksave-rtk.js"));
                let _ = fs::remove_dir_all(p.plugins_dir.join("toksave-rtk"));
                Ok(true)
            }
            ToolId::Ponytail => {
                remove_owner("opencode", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                remove_owner("opencode", "principles")?;
                Ok(true)
            }
        }
    }

    fn verify(&self, tool: ToolId) -> Option<bool> {
        let p = opencode_paths();
        let cfg = read_json_file(&p.config).ok().flatten();
        match tool {
            ToolId::Codegraph => Some(cfg.as_ref().is_some_and(|c| {
                crate::util::mcp::json_tool_healthy(
                    c,
                    "mcp",
                    crate::registry::AgentId::Opencode,
                    ToolId::Codegraph,
                ) || c
                    .get("mcp")
                    .and_then(|m| m.get("servers"))
                    .is_some_and(|s| {
                        let mut wrapper = serde_json::json!({});
                        wrapper["servers"] = s.clone();
                        crate::util::mcp::json_tool_healthy(
                            &wrapper,
                            "servers",
                            crate::registry::AgentId::Opencode,
                            ToolId::Codegraph,
                        )
                    })
            })),
            ToolId::ContextMode => Some(
                cfg.as_ref()
                    .map(|c| {
                        let in_v1 = c
                            .get("plugin")
                            .and_then(|p| p.as_array())
                            .map(|arr| arr.contains(&serde_json::json!("context-mode")))
                            .unwrap_or(false);
                        let in_v2 = c
                            .get("plugins")
                            .and_then(|p| p.as_array())
                            .map(|arr| {
                                arr.iter().any(|pl| {
                                    pl.as_str()
                                        .map(|s| s.contains("context-mode"))
                                        .unwrap_or(false)
                                })
                            })
                            .unwrap_or(false);
                        let in_mcp = c
                            .get("mcp")
                            .and_then(|m| m.get("servers"))
                            .and_then(|s| s.get("context-mode"))
                            .is_some();
                        in_v1 || in_v2 || in_mcp
                    })
                    .unwrap_or(false),
            ),
            ToolId::Caveman => Some(has_owner("opencode", "caveman")),
            ToolId::Rtk => Some(
                p.plugins_dir.join("toksave-rtk.js").exists()
                    || p.plugins_dir.join("toksave-rtk").exists(),
            ),
            ToolId::Ponytail => Some(has_owner("opencode", "ponytail")),
            ToolId::Principles => Some(has_owner("opencode", "principles")),
        }
    }
}
